zstash/utils.py

```diff
@@ -72,49 +74,107 @@ def run_command(command: str, error_str: str):
         raise RuntimeError(error_str)
 
 
-def get_files_to_archive(cache: str, include: str, exclude: str) -> List[str]:
+def get_files_to_archive_with_stats(
+    cache: str, include: str, exclude: str
+) -> Dict[str, Tuple[int, datetime]]:
+    """
+    OPTIMIZED VERSION: Gather list of files to archive along with their stats.
+
+    Uses os.scandir() to get file stats during the directory walk,
+    eliminating the need to stat files again later during database comparison.
+
+    Returns:
+        Dictionary mapping file_path -> (size, mtime)
+    """
     # PERFORMANCE: Start timing file gathering
     gather_total_start = time.time()
     logger.info("-" * 80)
-    logger.info("PERFORMANCE (get_files_to_archive): Starting file discovery")
+    logger.info(
+        "PERFORMANCE (get_files_to_archive_with_stats): Starting file discovery with stats"
+    )
 
-    # List of files
-    logger.info("Gathering list of files to archive")
+    # List of files with their stats
+    logger.info("Gathering list of files to archive (with stats)")
 
-    # PERFORMANCE: Time the os.walk operation
+    # PERFORMANCE: Time the os.scandir operation
     walk_start = time.time()
-    # Tuples of the form (path, filename)
-    file_tuples: List[Tuple[str, str]] = []
+    # Dictionary mapping path -> (size, mtime)
+    file_stats: Dict[str, Tuple[int, datetime]] = {}
     dir_count = 0
     file_count = 0
     empty_dir_count = 0
+    cache_path = os.path.join(".", cache)
 
-    # Walk the current directory
-    for root, dirnames, filenames in os.walk("."):
-        dir_count += 1
+    def scan_directory(path: str):
+        """Recursively scan directory using os.scandir() for efficiency."""
+        nonlocal dir_count, file_count, empty_dir_count
 
-        if not dirnames and not filenames:
-            # There are no subdirectories nor are there files.
-            # This directory is empty.
-            file_tuples.append((root, ""))
+        try:
+            entries = list(os.scandir(path))
+        except PermissionError:
+            logger.warning(f"Permission denied: {path}")
+            return
+
+        dir_count += 1
+        has_contents = False
+
+        for entry in entries:
+            # Skip the cache directory entirely
+            if entry.path == cache_path or entry.path.startswith(cache_path + os.sep):
```

Comment:
Is this going to work even if a user has a bad idea to create a cache like directory lower down in their files?